### ISSUE TITLE
Revise showing ETA and rates for progress bars

### ISSUE THREAD

**Issue by seqre** [praetorian-inc/noseyparker#140]
The progress bar [comment mentions an issue](https://github.com/praetorian-inc/noseyparker/blob/main/crates/progress/src/lib.rs#L8-L14) with the `indicatif` estimation algorithm for ETA and the throughput. The [PR linked in the comment](https://github.com/console-rs/indicatif/issues/394) with the fix for the issue was merged. It might be worth revisiting if ETA and rates can be reenabled.

**Comment by bradlarsen** [comment]
Yes, good idea! It would be more user-friendly to give meaningful ETAs and show current scan rate.

**Comment by bradlarsen** [comment]
Longer-term, I've been thinking about switching away from Indicatif entirely and using something lighter weight, like maybe [`status_line`](https://docs.rs/status-line/latest/status_line/index.html). I had noticed earlier on in the development of Nosey Parker that in parallel scan jobs, there was noticeable overhead just from updating the indicatif progress bar, and so Nosey Parker has resorted to some [dirty hacks](https://github.com/praetorian-inc/noseyparker/blob/abee5a81097e691a4c51efe56e99ed7f61187c32/crates/progress/src/lib.rs#L18-L29) that to minimize that.

**Comment by bradlarsen** [comment]
There is also an issue now with an interaction between the `log`, `tracing`, and `indicatif` crates: if a progress bar is active, any log messages will mess up how it is rendered. Probably the right way to fix that is by somehow modifying the [global `noseyparker` log setup code](https://github.com/praetorian-inc/noseyparker/blob/abee5a81097e691a4c51efe56e99ed7f61187c32/crates/noseyparker-cli/src/main.rs#L22-L60). Log messages shouldn't interfere with any active progress bar, and this shouldn't require modifying every single call site of `log` or `tracing` functions.

**Comment by bradlarsen** [comment]
Also longer-term, I think Nosey Parker will need to move away from its current scanning operation of (1) enumerating inputs to scan to determine progress bar maximum and then (2) scanning all those inputs.

This current mode of scan operation is problematic for a few reasons:

1. It requires two accesses of the filesystem, and there is time between them in which the filesystem could change. The total disk IO might be close to 2x as high as is necessary
2. It induces a blocking delay before scanning can begin: all the inputs need to be counted
3. It assumes that you can fully enumerate the inputs! This might not be feasible in the case of huge filesystem shares, for example. In such cases it would be beneficial if Nosey Parker could still do a partial scan.
4. It makes it difficult to implement an alternative scanning setup, such as a hypothetical Nosey Parker "server" mode where it would stay running until shut down, endlessly fulfilling requests to scan inputs sent to it

Without the "count all the inputs in advance" step, it may be impossible to provide an ETA, as Nosey Parker wouldn't have an idea how many inputs remained until it had actually scanned everything. (In this scenario, scan throughput rate and a total count of things scanned could still be reported, however.)

**Comment by seqre** [comment]
> I had noticed earlier on in the development of Nosey Parker that in parallel scan jobs, there was noticeable overhead just from updating the indicatif progress bar, and so Nosey Parker has resorted to some [dirty hacks](https://github.com/praetorian-inc/noseyparker/blob/abee5a81097e691a4c51efe56e99ed7f61187c32/crates/progress/src/lib.rs#L18-L29) that to minimize that.

I looked at indicatif documentation, wouldn't [`enable_steady_tick`](https://docs.rs/indicatif/latest/indicatif/struct.ProgressBar.html#method.enable_steady_tick) alleviate that issue? If [it's enabled](https://docs.rs/indicatif/latest/src/indicatif/progress_bar.rs.html#165-187), the normal operations such as `inc` [won't progress the bar](https://docs.rs/indicatif/latest/src/indicatif/progress_bar.rs.html#210-215).
